README.md
<a name="readme-top"></a>

<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://varadise.ai/">
    <img src="docs/images/logo.webp" alt="Logo" />
  </a>
  <h3 align="center">TCP2MQTT</h3>
  <p align="center">
    <br />

  </p>
</div>

<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="LICENSE.txt">License</a></li>
    <li><a href="#maintainers">Maintainers</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->
## About The Project

A proxy to recieve raw TCP sockets and send to a MQTT broker with customable format.
Froked from [tcp2mqtt](https://github.com/gonzalo123/tcp2mqtt)

It is a go client that reads the TCP sockets and send the information to the MQTT broker.
It support json format as TCP sockets foramt only.


<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- GETTING STARTED -->
## Getting Started

1. Edit the coustomated schema in config.yaml file
	```yaml
	version: 1
	mqtt:
	schema:
		message: |
		{{. | fjson}}
		username: |
		{{.IMEI | printf "%.f" -}}
		password: ""
	```
2. Enter the MQTT broker configuration in environment.BROKER in docker-compose.yml
```yaml
environment:
	- CONFIG_PATH=/opt/config.yaml
	- BROKER=tcp://localhost:1883
```
run ```docker-compose up --build```

### Prerequisites


### Installation


<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- USAGE EXAMPLES -->
## Usage

We use go-template to generate the MQTT message.
There is a customated function imported into the template engine.

```go
// orderedMarshalString marshals a given value into JSON with ordered keys
func orderedMarshalString(v any) (string, error) {
	b, err := encoder.Encode(v, encoder.SortMapKeys)
	if err != nil {
		return "", err
	}
	return string(b), nil
}
```
``` go
template.FuncMap{
	"fjson": orderedMarshalString,
}
```

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- MAINTAINERS -->
## Maintainers

An up-to-date list of people involved with development / support of the project

<!-- Simon Ball - [simonball@varadise.cloud](mailto:simonball@varadise.cloud) -->

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- MARKDOWN LINKS -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[product-screenshot]: docs/images/screenshot.png
<!-- Links to pages for products / components commonly used across the company -->
<!-- If you're project is using something not listed below, please create a ticket -->
<!-- in the Varadise Kitchen Board to propose updating the master template -->
[Next-url]: https://nextjs.org/
[React-url]: https://reactjs.org/
[Vue-url]: https://vuejs.org/
[Flask-url]: https://flask.palletsprojects.com
[Gin-url]: https://gin-gonic.com/

CHANGELOG.md
# Changelog

## 1.0.0

### Added or Changed
- Added this changelog :)
- Fixed typos in both templates
- Back to top links
- Added more "Built With" frameworks/libraries
- Changed table of contents to start collapsed
- Added checkboxes for major features on roadmap

### Removed

- Some packages/libraries from acknowledgements I no longer use

cmd/app/main_test.go
package main

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestHandleMsg(t *testing.T) {
	ip := "127.0.0.1"
	port := "7777"
	schema := []byte(`{"{{.netData.IMEI | printf "%.f"}}":[{"addr":"{{.netData.addr}}","cmd":"{{.netData.cmd}}","count":{{.netData.count}},"ipFrom":"{{.ip}}","port":"{{.port}}"}]}`)
	netData := (`{"addr":"1","cmd":"upload","count":[0,0,0,0,0,0,0],"IMEI":861714058319892}`)
	result, err := handleMsg(ip, port, netData, schema)
	assert.NoError(t, err)
	assert.Equal(t, (`{"861714058319892":[{"addr":"1","cmd":"upload","count":[0 0 0 0 0 0 0],"ipFrom":"127.0.0.1","port":"7777"}]}`), string(result))
	_, err = json.Marshal(result)
	assert.NoError(t, err)
}


pkg/config/config_test.go
package config

import (
	"os"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestGetMqttSchema(t *testing.T) {
	os.Setenv("CONFIG_PATH", "config.yaml")
	result, err := GetMqttSchema()
	assert.NoError(t, err)
	assert.Equal(t, string(`{"rawMsg":{{. | fjson}},"ipFrom":"{{.ip}}","port":"{{.port}}"}`), strings.TrimSpace(string(result)))
}

func TestGetUsernameSchema(t *testing.T) {
	os.Setenv("CONFIG_PATH", "config.yaml")
	result, err := GetUsernameSchema()
	assert.NoError(t, err)
	assert.Equal(t, string(`{{.IMEI | printf "%.f"}}`), strings.TrimSpace(string(result)))
}

func TestGetPasswordSchema(t *testing.T) {
	os.Setenv("CONFIG_PATH", "config.yaml")
	result, err := GetPasswordSchema()
	assert.NoError(t, err)
	assert.Equal(t, string(``), strings.TrimSpace(string(result)))
}

func TestGetMessageTypeSchema(t *testing.T) {
	os.Setenv("CONFIG_PATH", "config.yaml")
	result, err := GetMessageTypeSchema()
	assert.NoError(t, err)
	assert.Equal(t, string(`json`), strings.TrimSpace(string(result)))
}


pkg/message/message_test.go
package message

import (
	"encoding/json"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestHandleWholeMsg(t *testing.T) {
	ip := "127.0.0.1"
	port := "7777"
	msgSchema := []byte(`{{. | fjson}}`)
	usernameSchema := []byte(`{{.IMEI | printf "%.f"}}`)
	passwordSchema := []byte(``)
	netData := (`{"addr":"1","cmd":"upload","IMEI":861714058319892}`)
	messagetypeSchema := []byte(`json`)
	result, err := HandleWholeMsg(ip, port, netData, msgSchema, usernameSchema, passwordSchema, messagetypeSchema)
	assert.NoError(t, err)
	assert.Equal(t, (`{"IMEI":861714058319892,"addr":"1","cmd":"upload","ipFrom":"127.0.0.1","port":"7777"}`), string(result.Msg))
	assert.Equal(t, (`861714058319892`), string(result.Username))
	assert.Equal(t, (``), string(result.Password))
	_, err = json.Marshal(result)
	assert.NoError(t, err)

	// csv format
	msgSchema = []byte(`{{.| fjson}}`)
	usernameSchema = []byte(`{{.key5 | printf "%.f"}}`)
	passwordSchema = []byte(``)
	netData = (`1,2,3,a,4`)
	messagetypeSchema = []byte(`csv`)
	result, err = HandleWholeMsg(ip, port, netData, msgSchema, usernameSchema, passwordSchema, messagetypeSchema)
	assert.NoError(t, err)
	assert.Equal(t, (`{"ipFrom":"127.0.0.1","key1":1,"key2":2,"key3":3,"key4":"a","key5":4,"port":"7777"}`), string(result.Msg))
	assert.Equal(t, (`4`), string(result.Username))
	assert.Equal(t, (``), string(result.Password))
	_, err = json.Marshal(result)
	assert.NoError(t, err)
}

func TestHandleMsg(t *testing.T) {
	ip := "127.0.0.1"
	port := "7777"
	msgSchema := []byte(`{{. | fjson}}`)
	netData := (`{"addr":"1","cmd":"upload","IMEI":861714058319892}`)
	netDataJS := make(map[string]any)
	err := json.Unmarshal([]byte(strings.TrimSpace(netData)), &netDataJS)
	assert.NoError(t, err)
	result, err := handleMsg(ip, port, netDataJS, msgSchema)
	assert.NoError(t, err)
	assert.Equal(t, (`{"IMEI":861714058319892,"addr":"1","cmd":"upload","ipFrom":"127.0.0.1","port":"7777"}`), string(result))
	_, err = json.Marshal(result)
	assert.NoError(t, err)
}

func Test_csv2map(t *testing.T) {
	csvStr := `1,2,3,4,5`
	elementMap, err := csv2map(csvStr)
	assert.NoError(t, err)
	assert.Equal(t, (map[string]interface{}{"key1": 1., "key2": 2., "key3": 3., "key4": 4., "key5": 5.}), elementMap)

	// wrong msg format
	csvStr = "{1,2,3,4,5,6"
	elementMap, err = csv2map(csvStr)
	assert.Equal(t, (map[string]any{}), elementMap)
	assert.EqualErrorf(t, err, "csv2map: csvStr is not csv", "Error should be: %v, got: %v", "csv2map: csvStr is not csv", err)
}


pkg/config/config.go
package config

import (
	"os"

	"gopkg.in/yaml.v3"
)

func GetMqttSchema() ([]byte, error) {
	type V struct {
		MQTT struct {
			SCHEMA struct {
				MESSAGE string
			}
		}
	}
	y, err := os.ReadFile(os.Getenv("CONFIG_PATH"))
	if err != nil {
		return nil, err
	}
	v := V{}
	// m := make(map[string]interface{})
	err = yaml.Unmarshal([]byte(y), &v)
	if err != nil {
		return nil, err
	}
	return []byte(v.MQTT.SCHEMA.MESSAGE), nil
}

func GetUsernameSchema() ([]byte, error) {
	type V struct {
		MQTT struct {
			SCHEMA struct {
				USERNAME string
			}
		}
	}
	y, err := os.ReadFile(os.Getenv("CONFIG_PATH"))
	if err != nil {
		return nil, err
	}
	v := V{}
	// m := make(map[string]interface{})
	err = yaml.Unmarshal([]byte(y), &v)
	if err != nil {
		return nil, err
	}
	return []byte(v.MQTT.SCHEMA.USERNAME), nil
}

func GetPasswordSchema() ([]byte, error) {
	type V struct {
		MQTT struct {
			SCHEMA struct {
				PASSWORD string
			}
		}
	}
	y, err := os.ReadFile(os.Getenv("CONFIG_PATH"))
	if err != nil {
		return nil, err
	}
	v := V{}
	// m := make(map[string]interface{})
	err = yaml.Unmarshal([]byte(y), &v)
	if err != nil {
		return nil, err
	}
	return []byte(v.MQTT.SCHEMA.PASSWORD), nil
}

func GetMessageTypeSchema() ([]byte, error) {
	type V struct {
		MQTT struct {
			SCHEMA struct {
				MESSAGETYPE string
			}
		}
	}
	y, err := os.ReadFile(os.Getenv("CONFIG_PATH"))
	if err != nil {
		return nil, err
	}
	v := V{}
	// m := make(map[string]interface{})
	err = yaml.Unmarshal([]byte(y), &v)
	if err != nil {
		return nil, err
	}
	return []byte(v.MQTT.SCHEMA.MESSAGETYPE), nil
}


pkg/message/message.go
package message

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"tcp2mqtt/pkg/template"
)

type HandledMessage struct {
	Msg      string
	Username string
	Password string
}

// use template to hangle username
func handleUsername(bytesRepresentation []byte, schema []byte) ([]byte, error) {
	result, err := template.ApplyTemplate(schema, bytesRepresentation)
	if err != nil {
		return nil, err
	}
	return result, nil
}

func clear[M ~map[K]V, K comparable, V any](m M) {
    for k := range m {
        delete(m, k)
    }
}

type csvError struct {
	s string
}

func (e *csvError) Error() string {
	return e.s
}

// csv2map converts csv to map
func csv2map(csvStr string) (map[string]any, error) {
	elementMap := make(map[string]any)
	defer clear(elementMap)
	if csvStr[0] == '{' {
		err := &csvError{"csv2map: csvStr is not csv"}
		return map[string]any{}, err
	}
	s := strings.Split(csvStr, ",")
	for i := 0; i < len(s); i += 1 {
		num, err := strconv.ParseFloat(s[i], 64)
		if err != nil {
			elementMap[fmt.Sprintf("key%d", i+1)] = s[i]
		} else {
			elementMap[fmt.Sprintf("key%d", i+1)] = &num
		}
	}
	return elementMap, nil
}

// use template to hangle password
func handlePassword(bytesRepresentation []byte, schema []byte) ([]byte, error) {
	result, err := template.ApplyTemplate(schema, bytesRepresentation)
	if err != nil {
		return nil, err
	}
	return result, nil
}

// use template to hangle msg
func handleMsg(ip string, port string, bytesRepresentation []byte, schema []byte) ([]byte, error) {
	handledNetData, err := template.ApplyTemplate(schema, bytesRepresentation)
	if err != nil {
		return nil, err
	}
	msgData := make(map[string]interface{})
	defer clear(msgData)
	err = json.Unmarshal(handledNetData, &msgData)
	if err != nil {
		return nil, err
	}
	msgData["ipFrom"] = ip
	msgData["port"] = port
	return json.Marshal(msgData)
}

func HandleWholeMsg(ip string, port string, netData string, msgSchema []byte, usernameSchema []byte, passwordSchema []byte, messagetypeSchema []byte) (HandledMessage, error) {
	netDataJS := make(map[string]any)
	defer clear(netDataJS)
	var err error
	nilHandledMessage := HandledMessage{}
	if string(messagetypeSchema) == "csv" {
		netDataJS, err = csv2map(netData)
		if err != nil {
			return nilHandledMessage, err
		}
	} else {
		err := json.Unmarshal([]byte(strings.TrimSpace(netData)), &netDataJS)
		if err != nil {
			return nilHandledMessage, err
		}
	}
	bytesRepresentation, err := json.Marshal(netDataJS)
	if err != nil {
		return nilHandledMessage, err
	}
	msg, err := handleMsg(ip, port, bytesRepresentation, msgSchema)
	if err != nil {
		return nilHandledMessage, err
	}
	username, err := handleUsername(bytesRepresentation, usernameSchema)
	if err != nil {
		return nilHandledMessage, err
	}
	password, err := handlePassword(bytesRepresentation, passwordSchema)
	if err != nil {
		return nilHandledMessage, err
	}
	return HandledMessage{string(msg), string(username), string(password)}, nil

}


pkg/template/template.go
package template

import (
	"bytes"
	"text/template"

	"github.com/bytedance/sonic"
	"github.com/bytedance/sonic/encoder"
)

// orderedMarshalString marshals a given value into JSON with ordered keys
func orderedMarshalString(v any) (string, error) {
	b, err := encoder.Encode(v, encoder.SortMapKeys)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// ApplyTemplate gets template of an TCP socket package and applies it
func ApplyTemplate(templateData, data []byte) ([]byte, error) {
	t, err := template.New("template").Funcs(
		template.FuncMap{
			"fjson": orderedMarshalString,
		},
	).Parse(string(templateData))
	if err != nil {
		return nil, err
	}

	buf := new(bytes.Buffer)

	dataMap := make(map[string]any)
	if err := sonic.Unmarshal(data, &dataMap); err != nil {
		return nil, err
	}
	if err := t.Execute(buf, dataMap); err != nil {
		return nil, err
	}

	return buf.Bytes(), nil
}


cmd/app/main.go
package main

import (
	"bufio"
	"flag"
	"log"
	"net"
	"runtime"
	"tcp2mqtt/pkg/config"
	"tcp2mqtt/pkg/message"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

func main() {
	port, closeConnection, topic, broker := parseFlags()
	openSocket(*port, *closeConnection, *topic, *broker, onMessage)
}

func openSocket(port string, closeConnection bool, topic string, broker string, onMessage func(url string, username string, password string, topic string, buffer string)) {
	msgSchema, err := config.GetMqttSchema()
	if err != nil {
		log.Fatalln(err)
	}
	usernameSchema, err := config.GetUsernameSchema()
	if err != nil {
		log.Fatalln(err)
	}
	passwordSchema, err := config.GetPasswordSchema()
	if err != nil {
		log.Fatalln(err)
	}
	messagetypeSchema, err := config.GetMessageTypeSchema()
	if err != nil {
		log.Fatalln(err)
	}
	PORT := "0.0.0.0:" + port
	l, err := net.Listen("tcp4", PORT)
	log.Printf("Serving %s\n", l.Addr().String())
	if err != nil {
		log.Fatalln(err)
	}
	defer l.Close()

	for {
		c, err := l.Accept()
		if err != nil {
			log.Println(err)
			return
		}
		go handleConnection(c, msgSchema, usernameSchema, passwordSchema, messagetypeSchema, closeConnection, topic, broker, onMessage)
	}
}

func createClientOptions(url string, username string, password string) *mqtt.ClientOptions {
	opts := mqtt.NewClientOptions()
	opts.AddBroker(url)
	opts.SetUsername(username)
	opts.SetPassword(password)
	return opts
}

func connect(url string, username string, password string) mqtt.Client {
	opts := createClientOptions(url, username, password)
	client := mqtt.NewClient(opts)
	token := client.Connect()
	for !token.WaitTimeout(3 * time.Second) {
	}
	if err := token.Error(); err != nil {
		log.Fatal(err)
	}
	return client
}

func onMessage(url string, username string, password string, topic string, buffer string) {
	client := connect(url, username, password)
	token := client.Publish(topic, 2, false, buffer)
	log.Println(token)
	defer client.Disconnect(250)
}

func parseFlags() (*string, *bool, *string, *string) {
	port := flag.String("port", "7777", "port number")
	closeConnection := flag.Bool("close", true, "Close connection")
	topic := flag.String("topic", "topic", "mqtt topic")
	broker := flag.String("broker", "tcp://localhost:1883", "mqtt server")
	flag.Parse()

	return port, closeConnection, topic, broker
}

func handleConnection(c net.Conn, msgSchema []byte, usernameSchema []byte, passwordSchema []byte, messagetypeSchema []byte, closeConnection bool, topic string, broker string, onMessage func(url string, username string, password string, topic string, buffer string)) {
	log.Printf("Accepted connection from %s\n", c.RemoteAddr().String())
	for {
		ip, port, err := net.SplitHostPort(c.RemoteAddr().String())
		if err != nil {
			log.Println(err)
		}
		netData, err := bufio.NewReader(c).ReadString('}')
		if err != nil {
			log.Printf("error reading netData: %v", err)
		}
		log.Printf("netData: %v", netData)
		handledMessage, err := message.HandleWholeMsg(ip, port, netData, msgSchema, usernameSchema, passwordSchema, messagetypeSchema)
		if err != nil {
			log.Println(err)
		} else {
			log.Printf("sending to topic %s message:%s\n", topic, handledMessage.Msg)
			onMessage(broker, handledMessage.Username, handledMessage.Password, topic, handledMessage.Msg)
			log.Printf("sent")
		}
		if closeConnection {
			log.Printf("Closing connection from %s\n", c.RemoteAddr().String())
			c.Close()
			runtime.GC()
			return
		}
	}
}
